bgp router ID ip pool implementation

liyihuang opened the pull request:
Implement the Option1 in #30333


> Option1: Via Helm configuration
> 
> Provide Router ID pool through Cilium operator's configuration like bgpControlPlane.routerIDPool=10.0.0.0/16 which will ultimately rendered as a cilium-config ConfigMap entry. This is easy to implement, but changing or extending the range requires operator restart.
> 
> 

more details in commits messages 

Fixes: #30333
```release-note
Introduce a new ip pool mode for BGP router ID allocation so cilium can pick up the router ID from a pool.
```

liyihuang committed:
bgp: Add IP pool router-id allocation mode

Introduce a new BGP router-id allocation mode 'ip-pool' alongside the existing 'default' mode.

- Add a new configuration option for IP pool-based router-id allocation
- Modify CLI, Helm values, and configuration to support the new mode

Signed-off-by: Liyi Huang <liyi.huang@isovalent.com>

liyihuang committed:
bgp: Implement the ip-pool in operator and agent

- Add a new field for operator to allocate the ip pool
- Update the router-id for nodes when reconciling cluster
- Get the router ID in the agent
- Change rangeFromPrefix to be public so operator can use it

Signed-off-by: Liyi Huang <liyi.huang@isovalent.com>

harsimran-pabla commented on daemon/cmd/daemon_main.go:
```diff
@@ -992,9 +992,12 @@ func InitGlobalFlags(cmd *cobra.Command, vp *viper.Viper) {
 	option.BindEnv(vp, option.EnableBGPControlPlaneStatusReport)
 
-	flags.String(option.BGPRouterIDAllocationMode, defaults.BGPRouterIDAllocationMode, "BGP router-id allocation mode. Currently supported values: 'default' ")
+	flags.String(option.BGPRouterIDAllocationMode, option.BGPRouterIDAllocationModeNode, "BGP router-id allocation mode. Currently supported values: 'node' or 'ip-pool'")
```
I'm not big fan of changing helm flag, it causes unnecessary pain in upgrades. What is the issue with keeping it 'default' and 'ip-pool'? We can add some description to 'default' mode if it is not clear.

liyihuang replied:
Sure. I thought it's clearer to rename it but forgetting the upgrade.

harsimran-pabla commented on operator/pkg/bgpv2/cluster.go:
```diff
@@ -102,120 +119,164 @@ func (b *BGPResourceManager) upsertNodeConfigs(ctx context.Context, config *v2.C
-		matchingNodes.Insert(node.Name)
-
-		// Find node config for this node
+		bgpNodes := nodeSelector.Matches(slim_labels.Set(node.Labels))
```
Should be `bgpNode`, it's a single node match.

harsimran-pabla commented on operator/pkg/bgpv2/manager.go:
```diff
@@ -91,19 +97,37 @@ func registerBGPResourceManager(p BGPParams) *BGPResourceManager {
-	b.nodeConfigClient = b.clientset.CiliumV2().CiliumBGPNodeConfigs()
+		from, to := lbipam.RangeFromPrefix(ipnet)
+		// 50 router IDs are enough since we don't expect more than too many nodes to run BGP with upstream routers
+		b.bgpRouterIDIPPool, err = ipalloc.NewHashAllocator[string](from, to, 50)
```
Comment seems bit misleading, 50 is hash map init capacity not the limit on number of IPs.

harsimran-pabla commented on operator/pkg/bgpv2/cluster.go:
```diff
@@ -102,120 +119,164 @@ func (b *BGPResourceManager) upsertNodeConfigs(ctx context.Context, config *v2.C
+		if bgpNodes {
+			// Record selected node for later use
+			matchingNodes.Insert(node.Name)
+			if !oldNodeConfigExists && b.bgpRouterIDIPPool != nil {
```
qq: I see nil checks on bgpRouterIDIPPool, is there any case when bgpRouterIDIPPool can be nil?

harsimran-pabla replied:
I see, it is only set when router-id mode is enabled, consider using config option check instead of relying on b.bgpRouterIDIPPool to be nil/not nil for this condition.

liyihuang replied:
I dont know how to access the config without adding another field in the BGP manager. I felt that adding this `daemonConfig *option.DaemonConfig` to BGP manager is just too much, and that's why I use b.bgpRouterIDIPPool to determine. Shall I add `daemonConfig *option.DaemonConfig` to BGP manager?

harsimran-pabla replied:
> Shall I add daemonConfig *option.DaemonConfig to BGP manager?

That is reasonable, or just a flag that RouterIDPool is enabled.

harsimran-pabla commented on operator/pkg/bgpv2/cluster.go:
```diff
@@ -102,120 +119,164 @@ func (b *BGPResourceManager) upsertNodeConfigs(ctx context.Context, config *v2.C
+				if instance.RouterID == nil {
+					continue
+				}
+				routerID := netip.MustParseAddr(*instance.RouterID)
```
Use of MustParse should only be done in test, here we should do proper error handling.

harsimran-pabla commented on operator/pkg/bgpv2/cluster.go:
```diff
@@ -102,120 +119,164 @@ func (b *BGPResourceManager) upsertNodeConfigs(ctx context.Context, config *v2.C
+				}
+				routerID := netip.MustParseAddr(*instance.RouterID)
+				key := getRouterIDKey(node.Name, instance.Name)
+				_, exists := b.bgpRouterIDMap[key]
```
node might not belong to this cluster config, but to another cluster config. Are we incorrectly freeing up the router id?

liyihuang replied:
thanks for pointing this out. I dont think we need this logic anymore since the routerID free up are now part of `deleteNodeConfigs`. and the clean up when it's there is no config happenning in `if len(configs) == 0 && b.bgpRouterIDIPPool != nil `

harsimran-pabla commented on operator/pkg/bgpv2/cluster.go:
```diff
@@ -102,120 +119,164 @@ func (b *BGPResourceManager) upsertNodeConfigs(ctx context.Context, config *v2.C
-			// Conflict detected. Skip this node.
-			continue
 		}
+		if bgpNodes {
```
This whole code path could use less indentation, diff will be cleaner as well for review

```
if !bgpNodes { 
  continue
}
...
```

harsimran-pabla commented on operator/pkg/bgpv2/cluster.go:
```diff
@@ -102,120 +119,164 @@ func (b *BGPResourceManager) upsertNodeConfigs(ctx context.Context, config *v2.C
-				conflictingClusterConfigs.Insert(owner)
+		// restore or remove the router ID from the old NodeConfig if it exists
+		if oldNodeConfigExists && b.bgpRouterIDIPPool != nil {
+			for _, instance := range oldNodeConfig.Spec.BGPInstances {
```
Consider key cleanup/restore logic to be moved into another method.

liyihuang replied:
we dont have the cleanup logic here and I assume I should keep the codes here without another method. Is it correct?

harsimran-pabla commented on operator/pkg/bgpv2/manager.go:
```diff
@@ -91,19 +97,37 @@ func registerBGPResourceManager(p BGPParams) *BGPResourceManager {
+		b.bgpRouterIDIPPool, err = ipalloc.NewHashAllocator[string](from, to, 50)
+		if err != nil {
+			b.logger.Error("Failed to create BGP Router ID IP pool, BGP manager will not start", logfields.Error, err)
+			return nil
```
In error cases, set health state appropriately.

harsimran-pabla commented on operator/pkg/bgpv2/cluster.go:
```diff
@@ -102,120 +119,164 @@ func (b *BGPResourceManager) upsertNodeConfigs(ctx context.Context, config *v2.C
+				// because the oeprator could just get restarted
+				if bgpNodes && !exists {
+					b.bgpRouterIDMap[key] = ptr.To(routerID)
+					if err := b.bgpRouterIDIPPool.Alloc(routerID, key); err != nil {
```
Alloc and key storage should be swapped, otherwise in case of error internal state would already be updated.

harsimran-pabla reviewed (changes requested):
Thanks @liyihuang  for this contribution. Change is missing corresponding unit tests, please check out review comments and add the unit tests.

liyihuang committed:
bgp: Refactor tests to use DaemonConfig and add router ID allocation test

- Updated test fixtures to accept DaemonConfig instead of a boolean for status reporting.
- Introduced a new test for router ID allocation based on IP pool, validating multiple test cases.
- Enhanced existing tests to utilize the default DaemonConfig for consistency.

Signed-off-by: Liyi Huang <liyi.huang@isovalent.com>

squeed commented on pkg/option/config.go:
```diff
@@ -1074,9 +1074,12 @@ const (
 	BGPRouterIDAllocationMode = "bgp-router-id-allocation-mode"
 
+	// BGP router-id allocation IP pool
+	BGPRouterIDAllocationIPPool = "bgp-router-id-allocation-ip-pool"
```
Are these options used anywhere else? Is it possible to move them to a Cell in pkg/bgpv1?

liyihuang replied:
The operator is also using it. so it's not just pkg/bgpv1 cell. 

https://github.com/cilium/cilium/blob/48e010c5fc809d124397567e9b1210677622cb09/operator/pkg/bgpv2/manager.go#L112

The main logic happening in operator, agent is more for picking it up the values from the operator

harsimran-pabla replied:
I'd suggest looking at moving all BGP related flags to BGP operator cell.

thorn3r replied:
Agreed, moving them to a cell seems ideal

liyihuang replied:
I can have something like this in the operator cell.
```
type BGPConfig struct {
	EnableBGPControlPlane            
	EnableBGPControlPlaneStatusReport
	BGPRouterIDAllocationMode        
	BGPRouterIDAllocationIPPool      
	BGPSecretsNamespace              
}
```

But I can see we still this in the agent(https://github.com/cilium/cilium/pull/38300/commits/6c74bf307da66e7260f38ed722dbe9979f26df79#diff-23b500b5d2af6a22a607b920886922d961deb0e5f8c686f7481e4320722e017fR1226-R1241).

Do we want to just build a BGPconfig for operator cell? or I should consider to do that both for agent and operator?

liyihuang replied:
As discussed on slack. we will look into change the config into the cell in seperate PR

harsimran-pabla commented on install/kubernetes/cilium/values.yaml.tmpl:
```diff
@@ -488,8 +488,10 @@ bgpControlPlane:
   # -- BGP router-id allocation mode
   routerIDAllocation:
-    # -- BGP router-id allocation mode. In default mode, the router-id is derived from the IPv4 address if it is available, or else it is determined by the lower 32 bits of the MAC address.
+    # -- BGP router-id allocation mode. In node mode, the router-id is derived from the IPv4 address if it is available, or else it is determined by the lower 32 bits of the MAC address.
```
I thought we were reverting node->default naming. Same for commit message, please update that.

harsimran-pabla commented on operator/pkg/bgpv2/cluster.go:
```diff
@@ -102,31 +119,58 @@ func (b *BGPResourceManager) upsertNodeConfigs(ctx context.Context, config *v2.C
+		}
+		// Record selected node for later use
+		matchingNodes.Insert(node.Name)
+		if !oldNodeConfigExists && b.bgpRouterIDIPPoolEnabled {
```
can you add a comment why is this being done. I suppose it is for allocating new router id to this instance. Just for clarity.

liyihuang replied:
After your comments, I think I missed the scenario where the node exsits but the instance is newly added.

harsimran-pabla commented on operator/pkg/bgpv2/cluster.go:
```diff
@@ -102,31 +119,58 @@ func (b *BGPResourceManager) upsertNodeConfigs(ctx context.Context, config *v2.C
+		}
+		// Record selected node for later use
+		matchingNodes.Insert(node.Name)
+		if !oldNodeConfigExists && b.bgpRouterIDIPPoolEnabled {
```
This only runs if oldNodeConfigExists is false. How do we allocate router ids' to instances which are iteratively added to existing node config. 

- First config : NodeConfig does not exist, we will allocate router IDs from this code block.
- New instance added: Node config exists, but this code block is skipped. No router ID gets allocated to it.

harsimran-pabla commented on operator/pkg/bgpv2/cluster.go:
```diff
@@ -198,24 +242,38 @@ func (b *BGPResourceManager) upsertNodeConfigs(ctx context.Context, config *v2.C
-			if k8s_errors.IsNotFound(err) {
+		deleteErr := b.nodeConfigClient.Delete(ctx, nodeConfig.Name, meta_v1.DeleteOptions{})
+		// free the router ID from the IP pool and remove it from the map
+		if b.bgpRouterIDIPPoolEnabled {
```
I'd suggest moving this logic after error checking that API server returned no/idempotent error. So, we are not prematurely doing the cleanup.

harsimran-pabla commented on operator/pkg/bgpv2/cluster_test.go:
```diff
@@ -1220,6 +1234,124 @@ func TestDisableClusterConfigStatusReport(t *testing.T) {
+		nodeOverride     *v2.CiliumBGPNodeConfigOverride
+	}{
+		{
+			name:             "valid IPv4 router ID within pool",
```
I'd suggest adding few more test cases 
- Multiple Nodes
- Multiple Cluster configs
- Multiple instances per cluster config
- Test cleanup logic  with removal of instance, node, and cluster config.

harsimran-pabla replied:
Since router ID pool and management is cluster wide, we should test for all those combinations. The logic for managing router-ids is somewhat complex, it will be nice to get confidence in that logic with test cases which cover most if not all the corner cases.

tommyp1ckles commented on Documentation/helm-values.rst:
```diff
@@ -271,17 +271,21 @@
+     - ``""``
    * - :spelling:ignore:`bgpControlPlane.routerIDAllocation.mode`
-     - BGP router-id allocation mode. In default mode, the router-id is derived from the IPv4 address if it is available, or else it is determined by the lower 32 bits of the MAC address.
+     - BGP router-id allocation mode. In node mode, the router-id is derived from the IPv4 address if it is available, or else it is determined by the lower 32 bits of the MAC address.
```
Is node mode equivalent to "default" mode?

liyihuang replied:
yes. I tried to change the name but forgot to upodate this one

tommyp1ckles commented on operator/pkg/bgpv2/cluster.go:
```diff
@@ -102,31 +119,58 @@ func (b *BGPResourceManager) upsertNodeConfigs(ctx context.Context, config *v2.C
+				}
+				routerID, err := netip.ParseAddr(*instance.RouterID)
+				if err != nil {
+					errs = errors.Join(errs, err)
```
nit: might be nice to wrap some of these errors, if we end up getting a bunch of joined errors it will be hard to distinguish to what part of this procedure the failure was related to.

tommyp1ckles commented on operator/pkg/bgpv2/cluster.go:
```diff
@@ -347,6 +412,70 @@ func toNodeBGPInstance(clusterBGPInstances []v2.CiliumBGPInstance, overrideBGPIn
+	delete(b.bgpRouterIDMap, key)
+
+	if b.bgpRouterIDIPPool == nil {
+		return fmt.Errorf("bgp Router ID pool doesn't not exist")
```
nit: weird capitalization in `bgp Router` -> `BGP router`?

tommyp1ckles commented on operator/pkg/bgpv2/manager.go:
```diff
@@ -91,19 +99,42 @@ func registerBGPResourceManager(p BGPParams) *BGPResourceManager {
+	if p.DaemonConfig.BGPRouterIDAllocationMode == option.BGPRouterIDAllocationModeIPPool {
+		ipnet, err := netip.ParsePrefix(p.DaemonConfig.BGPRouterIDAllocationIPPool)
+		if err != nil {
+			b.logger.Error(err.Error())
```
nit: Probably should write a specific error message and then attach the err to the log, to provide information exactly what the origin of this log is.

tommyp1ckles commented on operator/pkg/bgpv2/manager.go:
```diff
@@ -91,19 +99,42 @@ func registerBGPResourceManager(p BGPParams) *BGPResourceManager {
+		}
+		if !ipnet.Addr().Is4() {
+			err = fmt.Errorf("BGP router ID IP pool must be an IPv4 address")
+			b.logger.Error(err.Error())
```
Same comment re: error logging, as above.

tommyp1ckles commented on operator/pkg/bgpv2/manager.go:
```diff
@@ -91,19 +99,42 @@ func registerBGPResourceManager(p BGPParams) *BGPResourceManager {
+		ipnet, err := netip.ParsePrefix(p.DaemonConfig.BGPRouterIDAllocationIPPool)
+		if err != nil {
+			b.logger.Error(err.Error())
+			b.health.Degraded("BGP manager health degraded", err)
```
BGP being degraded is somewhat implied from the auto generated health reporter ID plus the fact that the reporter is set to degraded, perhaps we should provide some context on the degradation? (ex. "bad BGP router allocation pool").

liyihuang replied:
I have added more info on the err itself, It's err log and `BGP manager health degraded` with the err now

tommyp1ckles commented on operator/pkg/bgpv2/manager.go:
```diff
@@ -91,19 +99,42 @@ func registerBGPResourceManager(p BGPParams) *BGPResourceManager {
+		if err != nil {
+			b.logger.Error(err.Error())
+			b.health.Degraded("BGP manager health degraded", err)
+			return nil
```
I don't think that logging/health reporting then returning nil is the best way to handle errors in the constructor.  If we know that the type can't be constructed (i.e. due to malformed configuration flag cidr) then its probably best to just return an error alongside the `*BGPResourceManager` and have hive report the error on termination.

By returning nil we also risk creating confusing termination error messages that complain about missing dependencies rather than the actual issue at hand.

tommyp1ckles commented on operator/pkg/bgpv2/manager.go:
```diff
@@ -91,19 +99,42 @@ func registerBGPResourceManager(p BGPParams) *BGPResourceManager {
+			err = fmt.Errorf("BGP router ID IP pool must be an IPv4 address")
+			b.logger.Error(err.Error())
+			b.health.Degraded("BGP manager health degraded", err)
+			return nil
```
Same issue as described above (re: failed constructor).

tommyp1ckles commented on operator/pkg/bgpv2/manager.go:
```diff
@@ -91,19 +99,42 @@ func registerBGPResourceManager(p BGPParams) *BGPResourceManager {
+		if err != nil {
+			b.logger.Error(err.Error())
+			b.health.Degraded("BGP manager health degraded", err)
+			return nil
```
Same issue as described above (re: failed constructor).

thorn3r reviewed (approved):
agent changes lgtm, nice work! :+1:

harsimran-pabla reviewed (approved):
Thanks @liyihuang for this change, LGTM now.
